### PR TITLE
First implementation of the Hypnogram class

### PR DESCRIPTION
**Pull request by raphaelvallat**
Hi,

This PR introduces a first working implementation of the `yasa.Hypnogram` class, which will be the new standard way to deal with hypnograms in YASA moving forward, as discussed in #105.

This should greatly simplify with implementation of the performance evaluation pipeline (#78).

**Remaining tasks**

- [x] Unit tests for 3, 4 and 5-stages hypnogram (see [coverage report](https://app.codecov.io/gh/raphaelvallat/yasa/pull/116))
- [x] Add `Hypnogram.consolidate_stages` method
- [ ] ~Implement the `Hypnogram.plot_hypnogram` function. Here I'm not sure if we should re-implement from scratch, or instead use the existing `yasa.plot_hypnogram` function and add support to 2, 3 and 4-stages hypnograms. The latter will require adding the parameter `n_stages` to the function however.~ @remrama will implement in a separate PR.
- [ ] Add compatibility with other YASA functions (e.g. detection). For now, users can simply use `Hypnogram.as_int()` to get a NumPy array with integer values.

@remrama I would appreciate your review on this (but no rush!). Let me know if you see any ideas for improvements and/or new class methods or properties that may be useful.

Thanks,
Raphael

### CODE REVIEW

File: yasa/hypno.py
Comment (remrama) [inline yasa/hypno.py:515]: I'm happy to implement this. I do think it should be attached here as a method. I have some plans for potentially consolidating the current `plot_hypnogram` code, and I also need to make some mods to plot multiple lines for the Evaluation PR (to overlay true/pred). So I think you could remove it from the TODO list on this PR, or if you want to add it here I can copy/paste into this thread the code I wrote for it.

File: yasa/hypno.py
Comment (remrama) [inline yasa/hypno.py:517]: I could see this function being useful outside of this class, maybe as a utility in some cases. Do you think it should be left out and then just call it here from the class? Or maybe it's cleaner to just have users make a Hypnogram class quickly to use it. Cool with me.

File: yasa/tests/test_hypnoclass.py
Comment (remrama) [inline yasa/tests/test_hypnoclass.py:27]: is printing useful in tests?

**Review by remrama — COMMENTED** [review]
SO COOL @raphaelvallat! No major suggestions from me. As usual, lots to say but not much of substance 🙃

File: yasa/hypno.py
Comment (remrama) [reply yasa/hypno.py:515]: Oh sorry, to clarify, I think this plotting function should remain in the separate plotting module and then called from here. I think it's useful to have separately, and also keeps all the plt imports and such in an isolated module.

File: yasa/hypno.py
Comment (remrama) [inline yasa/hypno.py:133]: Since we are passing strings into this class, we should probably switch it up so `yasa.simulate_hypno` returns strings (to avoid this step in many places).

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:32]: Good point!

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:43]: I think my preference is for loose validation, as long as it's clear to the user that YASA will do some internal operations to homogenize the stages. As suggested I've added two sentences about the abbreviation and lower/upper case — let me know if that's (not) enough!

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:58]: Yep, edited!

File: yasa/hypno.py
Comment (remrama) [reply yasa/hypno.py:232]: Let me offer a clarification here, because I think I misunderstood what was going on here at first. I've been playing around with this a bit since. I think it would be nice to have it set up like this:

* `Hypnogram.hypno` = The pandas.Series
* `Hypnogram.values` = numpy array of stages as integers
* `Hypnogram.names` = numpy array of stages as strings ("names" or something similar)
* `Hypnogram.stages` = list of possible labels (currently `.labels`)
* `Hypnogram.mapping_str2int` and `Hypnogram.mapping_int2str` for the two respective mapping dictionaries.

This would replace `Hypnogram.as_int()` and `.mapping`. The idea would be that there is no "as something" because there is no assumption that the hypnogram is normally one way and then converted to another. It's like the Hypnogram is an all-encompassing thing that contains lots of info (integer format, string format, scorer etc.) and you can pull anything out directly.

Thoughts?

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:134]: Yeah that's a great idea! I would do it in a future PR however

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:177]: I was too lazy to write them. Honestly I think these are pretty much self-explanatory, no?

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:62]: Good idea!

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:264]: Would it help to say: "The number of allowed stages in the hypnogram. This is not the number of unique stages in the current hypnogram."

File: yasa/hypno.py
Comment (remrama) [reply yasa/hypno.py:177]: Not a huge deal. Yes, definitely self-explanatory to those familiar with reading assertion errors. I'm only thinking of those who are using YASA as their first Python experience (which I think is common for YASA).

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:264]: I'd like to keep the attributes names concise (i.e. maximum 1 underscore 😁 )

Comment (remrama) [reply yasa/hypno.py:264]: Agreed about brevity and 1 underscore is a nice rule :)  Ya probably just clarify w/ that sentence. If anyone wants to get the number of unique in the hypnogram they can easily with `.nunique()` on the Series 👍

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:232]: This is quite a major change so let me think about it. A few first comments:

- Currently, users can define a custom mapping with `hyp.mapping = {}`, which is a nice feature. How would you implement that with this new structure?
- I am worried about having too many class attributes — the simpler, the better. What about `values` for strings and `values_int` for integers?
- I can rename `labels` to `stages` if that makes more sense

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:313]: It's supported via the `@mapping.setter` property. My thought was that:

1. For most users, asking them to instantiate the class with both the string stage values AND the mapping is going to be too much. Plus, I'd like to avoid having too many parameters in the Hypnogram class.
2. However for more advanced users there is still the (hidden) possibility to reset the mapping once and for all

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:358]: Nope, good catch!

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:391]: Nothing they are just going to be printed as is. I'm pretty sure I've seen that in other libraries. Happy to remove if that's more confusing than helpful!

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:515]: > this plotting function should remain in the separate plotting module and then called from here.

I agree

> I think you could remove it from the TODO list on this PR

Perfect

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:517]: I plan to keep the `yasa.sleep_statistics` function intact for now. Long term however I'd like all users to switch to the Hypnogram class method.

File: yasa/tests/test_hypnoclass.py
Comment (raphaelvallat) [reply yasa/tests/test_hypnoclass.py:27]: Needed for coverage

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:133]: I was hoping you'd say that :) We can implement this in a future PR though, no rush

File: yasa/hypno.py
Comment (remrama) [reply yasa/hypno.py:313]: Yes agree 100% this makes sense.

File: yasa/hypno.py
Comment (remrama) [reply yasa/hypno.py:391]: No no that's cool, I was really just curious.

File: yasa/hypno.py
Comment (remrama) [reply yasa/hypno.py:517]: Right, but then why not just use that function here? Maybe I'm missing a critical difference but at a quick glance it seems like this would just be repeating lots of the code?

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py]: Added!

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py]: Added, lmk!

File: yasa/hypno.py
Comment (remrama) [reply yasa/hypno.py:251]: Don't worry about this, leave it all as-is (including `labels`). I think I spoke too quickly, the current setup works great. For any advanced user, working with the `Hypnogram.hypno` Series solves any need for creativity. I've been working on re-implementing the Evaluation PR the last few hours and honestly this new class is _sooo_ nice.

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:543]: In this new implementation I'm really taking advantage of the numerous attributes (e.g. `.labels`, `.n_epochs`, `.sampling_frequency`) and methods (`.transition_matrix()`). I guess these could be passed to the `yasa.sleep_statistics` function but I thought it would be just easier to repeat some code rather than risking breaking the existing function. I can check next week how much work it would be to adapt the original function.

Comment (remrama) [reply yasa/hypno.py:543]: No, again don't stress over it. I'm sure you're taking the right approach, I was mostly just curious about your thought process (should've just asked that). This clarification makes sense 👍

**Review by remrama — COMMENTED** [review]
Current status approved in spirit, but I see you have more you'd like to implement so lmk when you want a full review with formal approval.

**Comment by remrama** [comment]
Btw, you're call but I would support merging this as-is and then adding "full functionality" in a separate PR. I'm not sure how big of a project that is, but if just the base class was in master I'd start some PRs with it.

**Comment by remrama** [comment]
Suggested method:

```python
class Hypnogram:
    ...
    def summary(self): # or .get_dataframe
        """
        Return a pandas DataFrame summarizing epoch-level information.

        Column order and names are compliant with BIDS events files [BIDSevents]_
        and MNE events/annotations dataframes [MNEannotations]_.

        Returns
        -------
        summary : :py:class:`pandas.DataFrame`
            A dataframe containing epoch onset, duration, stage, etc.

        References
        ----------
        .. [BIDSevents] https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/05-task-events.html
        .. [MNEannotations] https://mne.tools/stable/glossary.html#term-annotations
        """
        data = {
            "onset": self.timedelta.total_seconds(),
            "duration": 1 / self.sampling_frequency,
            "value": self.as_int().to_numpy(),
            "description": self.hypno.to_numpy(),
            "epoch": 1 + np.arange(self.n_epochs),
        }
        if hypno.scorer is not None:
            data["scorer"] = hypno.scorer
        return pd.DataFrame(data)
```

**Comment by raphaelvallat** [comment]
@remrama agreed for the new method! What do you think about calling it `Hypnogram.as_annotations()` or `Hypnogram.as_mne_annotations()`?

Also, when `start` is not None, should `onset` be the actual datetime, or is it better to always have `self.timedelta.total_seconds()`? I'm guessing the latter is the standard MNE/BIDS format?

I'll wait for your reply, add this new method and then request your final approval before merging 🎉 !

**Comment by remrama** [comment]
> What do you think about calling it `Hypnogram.as_annotations()` or `Hypnogram.as_mne_annotations()`?

Either are cool with me, although it's probably more of a BIDS/events emphasis than an MNE/annotation emphasis.

> Also, when `start` is not None, should `onset` be the actual datetime, or is it better to always have `self.timedelta.total_seconds()`? I'm guessing the latter is the standard MNE/BIDS format?

Ya good question. Definitely want to always have the `onset` column as seconds from start (ie, `self.timedelta.total_seconds()`), because this is the [BIDS-standard](https://bids-specification.readthedocs.io/en/stable/glossary.html#objects.columns.onset) which I think takes precedent. MNE is a bit more ambiguous (as far as I can tell), and sometimes returns onset as seconds or timestamps.

I think in this case, if the Hypnogram includes timestamp info it should be added as an _additional_ column rather than replacing `onset`. It's not totally clear to me what this would be called, according to BIDS docs, [this might be "sample"](https://bids-specification.readthedocs.io/en/stable/glossary.html#sample-columns) but I'm not sure. If unclear, I think we could just call it "timestamp".

**Comment by raphaelvallat** [comment]
In the future, maybe we could go even one step further and add an `output_type="mne"` parameter to this method, which could be either `"mne"` (returns a [mne.Annotations](https://mne.tools/stable/generated/mne.Annotations.html#mne.Annotations) object), or `"dataframe"` (default BIDS-like dataframe, which is also compatible with [EDFBrowser](https://www.teuniz.net/edfbrowser/EDFbrowser%20manual.html#Annotations)).

Question: why start `epoch` at 1 and not 0? 

```python
{"epoch": 1 + np.arange(self.n_epochs)}
```

> If unclear, I think we could just call it "timestamp".

After second thought, I think that for the initial implementation I would only include `onset` in seconds, and not the actual timestamps.

**Comment by raphaelvallat** [comment]
Maybe `epoch` could be set as the index of the resulting dataframe? Would that break things w.r.t to BIDS/MNE annotations format?

```python
>>> from yasa import Hypnogram
>>> hyp = Hypnogram(["W", "W", "LIGHT", "LIGHT", "DEEP", "REM", "WAKE"], n_stages=4)
>>> hyp.as_annotations()
       onset  duration  value description
epoch                                    
1        0.0      30.0      0        WAKE
2       30.0      30.0      0        WAKE
3       60.0      30.0      2       LIGHT
4       90.0      30.0      2       LIGHT
5      120.0      30.0      3        DEEP
6      150.0      30.0      4         REM
7      180.0      30.0      0        WAKE
```

**Comment by remrama** [comment]
Love the `output_type` idea!

> Question: why start epoch at 1 and not 0?

Ya you called me out on this :)  Start it at 0. I wasn't sure about that one, it's just that technically it _is_ the 1st epoch, and the Python indexing can be confusing for new users. But you're right, submit to Python indexing.

> I think that for the initial implementation I would only include `onset` in seconds, and not the actual timestamps.

Yep that's good. If someone really wants timestamp that can add it trivially with one more line.

> Maybe `epoch` could be set as the index of the resulting dataframe?

That looks great! It won't break things, neither BIDS or MNE expect that column. It's not an incredibly useful column anyways, it's really just a row number... Leave it starting at 0 and so basically we're just renaming that axis to `epoch`.

**Comment by raphaelvallat** [comment]
Ready for your final review @remrama !

File: yasa/hypno.py
Comment (remrama) [inline yasa/hypno.py:56]: typo "meanng" --> "meaning"

File: yasa/hypno.py
Comment (remrama) [inline yasa/hypno.py:727]: Sorry I missed this before. I suppose it makes sense to go ahead and include them all though I'm indifferent about it.

**Review by remrama — APPROVED** [review]
Ship it ✈️

File: yasa/hypno.py
Comment (raphaelvallat) [reply yasa/hypno.py:727]: Prefer not to include for now, as the output of this function is already quite dense

**Comment by raphaelvallat** [comment]
Merged to `master`. Thanks so much for your help on this! The next release is going to be 🔥

**Pull request merged**
